server/app/Icepeak/Main.hs

```diff
@@ -79,31 +79,52 @@ runCore core = do
 
   -- start threads
   pops <- Async.async $ Core.runCommandLoop core
-  sync <- Async.async $ Core.runSyncTimer core
   upds <- Async.async $ WebsocketServer.processUpdates core
   serv <- Async.async $ Server.runServer logger wsServer httpServer
+  sync <- Async.async (Core.runSyncTimer core)
   metrics <- Async.async
     $ forM_ (configMetricsEndpoint config) (MetricsServer.runMetricsServer logger)
+
   installHandlers core serv
   logAuthSettings config logger
   logQueueSettings config logger
   logSyncSettings config logger
   postLog logger "System online. ** robot sounds **"
 
-  waitLog "core" logger pops
-  waitLog "web sockets server" logger upds
-  waitLog "http server" logger serv
-  -- kill auxiliary threads when the main threads ended
-  Async.cancel metrics
-  Async.cancel sync
-
--- | Wait for an Async computation to exit and log unexpected exceptions.
-waitLog :: Text -> Logger -> Async () -> IO ()
-waitLog name logger action = Async.waitCatch action >>= handleLog where
-  handleLog (Left exc)
-    | Just (_ :: AsyncException) <- fromException exc = pure ()
-    | otherwise = Logger.postLog logger $ name <> " stopped with an exception: " <> Text.pack (show exc)
-  handleLog (Right _) = pure ()
+  runCoreResult <- waitFirstRunCoreResult pops upds serv
+  logRunCoreResult logger runCoreResult
+
+  -- kill all threads when one of the main threads ended
+  mapM_ Async.cancel [pops, upds, serv, metrics, sync]
+
+-- | Data type to hold results for the async that finishes first
+data RunCoreResult
+  = RCR_CommandLoop SomeException
+  | RCR_WebSockets SomeException
+  | RCR_HTTP SomeException
+  | RCR_Success
+
+-- | Wait on the first async to finish
+waitFirstRunCoreResult :: Async () -> Async () -> Async () -> IO RunCoreResult
+waitFirstRunCoreResult x y z = waitAllThree' `finally` mapM_ cancel [x, y, z]
```

Comment:
What does this do what `waitAnyCatchCancel` cannot do?

Maybe moving catching the exceptions into the async computations and having them return a status value is cleaner. But that's just an idea.
  
Also, the arguments seem to correspond to specific async actions (core, server, websockets), so they should be named accordingly.

Reply:
Difference is that it allows us to see which of the threads threw the error. `waitAnyCatchCancel` only returns a `SomeException`, so they all look the same regardless of which thread threw it.

I've changed it now in the way you mentioned, so that the catch and tagging of the exception happens within the async. Now i can simply use `waitAny` instead of the, admittedly kind of odd, `waitFirstRunCoreResult`.